spec/lib/whitehall_importer/change_history_spec.rb

```diff
@@ -0,0 +1,81 @@
+                             editions: [first_edition, superseded_edition, published_edition])
+
+    expect(described_class.new(whitehall_export).for(3)).to match([
+      { "id" => anything, "note" => "Some fantastic changes", "public_timestamp" => Time.zone.now.beginning_of_day.rfc3339 },
+    ])
+  end
+
+  it "returns draft edition change history if it has been previously unpublished" do
+    edition = build(:whitehall_export_edition,
+                    state: "draft",
+                    change_note: "Some great changes",
+                    unpublishing: build(:whitehall_export_unpublishing),
+                    revision_history: [build(:whitehall_export_revision_history_event,
+                                             state: "published",
+                                             created_at: Time.zone.now.beginning_of_day.rfc3339)])
+    whitehall_export = build(:whitehall_export_document,
+                             editions: [first_edition, edition])
+
+    expect(described_class.new(whitehall_export).for(3)).to match([
+      { "id" => anything, "note" => "Some great changes", "public_timestamp" => Time.zone.now.beginning_of_day.rfc3339 },
+    ])
+  end
+
+  it "doesn't return change history for minor change editions" do
+    edition = build(:whitehall_export_edition,
+                    state: "published",
+                    minor_change: true,
+                    change_note: "",
+                    revision_history: [build(:whitehall_export_revision_history_event, state: "published")])
+    whitehall_export = build(:whitehall_export_document,
+                             editions: [first_edition, edition])
+
+    expect(described_class.new(whitehall_export).for(3)).to be_empty
+  end
+
+  it "raises if an edition has a major change but no change note" do
+    edition = build(:whitehall_export_edition,
+                    state: "published",
+                    change_note: "",
+                    revision_history: [build(:whitehall_export_revision_history_event, state: "published")])
+    whitehall_export = build(:whitehall_export_document,
+                             editions: [first_edition, edition])
+
+    expect { described_class.new(whitehall_export).for(2) }
+      .to raise_error(WhitehallImporter::AbortImportError, "Edition has a major change but no change note")
+  end
+
+  it "raises if an edition has a major change but no publish event" do
+    edition = build(:whitehall_export_edition,
+                    state: "published",
+                    change_note: "Some change",
+                    revision_history: [build(:whitehall_export_revision_history_event)])
+    whitehall_export = build(:whitehall_export_document,
+                             editions: [first_edition, edition])
+
+    expect { described_class.new(whitehall_export).for(2) }
+      .to raise_error(WhitehallImporter::AbortImportError, "Edition has a major change but no publish event")
+  end
+end
```

Comment:
This test looks pretty good, a couple of cases still to cover. It does seem pretty heavy on the boilerplate needed for it. I wonder if we can reduce some of that with factory bot traits. I know we have one for published that could reduce some of the code here: https://github.com/alphagov/content-publisher/blob/bc6b81dead92587cebd65dba2d36fe27a95c4a88/spec/factories/whitehall_export/edition_factory.rb#L60-L75.

May also want one for superseded and to have published and superseded accept a published_at transient attribute.

Reply:
Sure, this cleans up the tests nicely, don't know why I didn't use the factories! 🤦‍♂